### ISSUE TITLE
[autoscaler] ResourceDemandScheduler legacy yaml logic

### ISSUE THREAD

**Issue by Gekho457** [ray-project/ray#11920]
### What is the problem?

The two blocks of logic in `ResourceDemandScheduler._infer_legacy_node_resources_if_needed()` look like they are intended to be executed once per monitoring/autoscaling session. However, the `ResourceDemandScheduler` is reinstantiated in `StandardAutoscaler.reset()`, which is called during each autoscaler `update`. As a result, the logic is repeated during each update when launching with a legacy-style yaml. This leads to an `AssertionError` at some point after the autoscaler starts a worker node. 

Here's the relevant part of the output from  `/tmp/ray/session_latest/logs/monitor.err` on the head node. 
```
Worker node types:
 - ray-legacy-worker-node-type: 1
2020-11-10 19:57:25,083 ERROR autoscaler.py:130 -- StandardAutoscaler: Error during autoscaling.
Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 128, in update
    self._update()
  File "/root/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 219, in _update
    ensure_min_cluster_size=self.resource_demand_vector)
  File "/root/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/resource_demand_scheduler.py", line 108, in get_nodes_to_launch
    self._infer_legacy_node_resources_if_needed(max_resources_by_ip)
  File "/root/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/resource_demand_scheduler.py", line 213, in _infer_legacy_node_resources_if_needed
    assert len(max_resources_by_ip) == 1  # Only the head node.
AssertionError
```

*Ray version and other system information (Python version, TensorFlow version, OS):*
Current master

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

I get this error when running `ray up python/ray/autoscaler/kubernetes/example-full.yaml` on a minikube cluster with the following modifications to the legacy-style `python/ray/autoscaler/kubernetes/example-full.yaml`:
head and worker container `image` fields set to `rayproject/ray:nightly`
`min_workers` set to 1.

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by Gekho457** [comment]
cc @wuisawesome @ericl

**Comment by AmeerHajAli** [comment]
Should be closed with #11802 .

**Comment by simon-mo** [comment]
The PR looks like a big PR. Can you push some hotfix so we can still use autoscaler master in the meantime?
